### PR TITLE
Mermaid front end rendering

### PR DESCRIPTION
**Pull request by charli117**
当生成结果符合mermaid代码块格式要求时，显示SVG转换图标，点击转换图标可在代码块与SVG矢量图间相互切换

### CODE REVIEW

**Comment by charli117** [comment]
**Promot可以参考如下：**
你将扮演mermaid图形生成器，根据用户场景描述生成符合mermaid格式要求的代码块。

[注意事项]
- 仅输出mermaid代码块，不做其他解释。

Let's think step by step.

**效果演示**

https://github.com/langgenius/dify/assets/24435100/3e44a291-a855-48ee-a91c-33e0074ace53

File: web/app/components/app/chat/mermaid/index.tsx [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    height: 'auto',
+    overflow: 'auto',
+  };
+
+  return (
+    <div>
+      {
+        isRender && <div id={chartId.current} className="mermaid" style={style}>{svgCode && (<img src={svgCode} style={{width: '100%', height: 'auto'}} alt="Mermaid chart" />)}</div>
+      }
+      <button onClick={handleReRender}>{t('appApi.merMaind.rerender')}</button>
```
Comment (crazywoola) [inline web/app/components/app/chat/mermaid/index.tsx:95]: Translation is missing here.

File: web/app/components/app/chat/mermaid/index.tsx [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+      useMaxWidth: true,
+    },
+  });
+  mermaidAPI = mermaid.mermaidAPI;
+}
+
+
+const Flowchart = React.forwardRef((props: {
+  PrimitiveCode: string;
+}, ref) => {
```
Comment (crazywoola) [inline web/app/components/app/chat/mermaid/index.tsx:30]: 'ref' is defined but never used.

File: web/app/components/base/markdown.tsx [hunk 3]
```diff
@@ -9,17 +9,28 @@ import { atelierHeathLight } from 'react-syntax-highlighter/dist/esm/styles/hljs
...
   css: 'CSS',
   html: 'HTML',
   xml: 'XML',
   php: 'PHP',
+  python: 'Python',
+  yaml: 'Yaml',
+  shell: 'Shell',
+  powershell: 'PowerShell',
+  markdown: 'MarkDown',
+  makefile: 'MarkFile',
```
Comment (crazywoola) [inline web/app/components/base/markdown.tsx]: typo? :)

File: web/app/components/base/markdown.tsx [hunk 4]
```diff
@@ -95,24 +107,36 @@ export function Markdown(props: { content: string }) {
...
-                      isPlain
-                    />
+                    <div style={{'display': 'flex'}}>
+                      {language === 'mermaid' &&
+                        <SVGBtn
+                          isSVG={isSVG}
+                          setIsSVG={setIsSVG}
+                        />
+                      }
+                      <CopyBtn
```
Comment (crazywoola) [inline web/app/components/base/markdown.tsx:116]: The outer box has a CopyBtn Already, so please consider remove this button.

File: web/app/components/app/chat/mermaid/index.tsx [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    const blob = new Blob([svgBytes], { type: 'image/svg+xml;charset=utf-8' });
+    return new Promise((resolve, reject) => {
+      const reader = new FileReader();
+      reader.onloadend = () => resolve(reader.result);
+      reader.onerror = reject;
+      reader.readAsDataURL(blob);
+    });
+  };
+
+  useEffect(() => {
```
Comment (crazywoola) [inline web/app/components/app/chat/mermaid/index.tsx]: Make sure useEffect is above render function. The handleReRender move before it.

**Review by crazywoola — CHANGES_REQUESTED** [review]

File: web/app/components/base/markdown.tsx (hunk 4 above)
Comment (charli117) [reply web/app/components/base/markdown.tsx:116]: 外壳原有的CopyBtn是实现代码块的复制，SVGBtn是切换源代码和SVG图像，功能层面本身并不冲突，而且只有在language为mermaid时才会显示SVGBtn

File: web/app/components/app/chat/mermaid/index.tsx (hunk 1 above)
Comment (charli117) [reply web/app/components/app/chat/mermaid/index.tsx:95]: 不好意思，这部分忘记合并了，立马补上提交

**Review by crazywoola — APPROVED** [review]
LGTM

**Pull request merged**